awpy/stats/win_prob.py

```diff
@@ -0,0 +1,87 @@
+"""Calculates CT & T side likeliehood of winning a round at any given point"""
+
+from typing import List, Dict, Union
+
+from awpy import Demo
+
+
+def build_feature_matrix(demo: Demo, ticks: Union[int, List[int]]) -> List[Dict[str, float]]:
+    """
+    Builds the game state feature matrix for specified ticks.
+
+    Args:
+        demo (Demo): A parsed demo object containing CS2 match data.
+        ticks (Union[int, List[int]]): A single tick or a list of ticks at which to calculate features.
+
+    Returns:
+        List[Dict[str, Union[int, str, bool]]]: A list of dictionaries where each dictionary contains the game state features at a given tick.
+    
+    Raises:
+        ValueError: If specified ticks are not found within the demo object.
+    """
+    if not demo.ticks.empty:
+        if isinstance(ticks, int):
+            ticks = [ticks]
+        
+        # Gather round numbers for each specified tick
+        round_numbers = []
+        for tick in ticks:
+            tick_data = demo.ticks[demo.ticks["tick"] == tick]
+            if tick_data.empty:
+                raise ValueError(f"No data found for tick {tick}.")
+            round_numbers.append(tick_data["round"].iloc[0])
+        
+        # Check if all ticks belong to the same round
+        if len(set(round_numbers)) > 1:
+            raise ValueError("Specified ticks are from different rounds. Please ensure all ticks belong to the same round.")
+            
+        game_state = []
+
+        # Process Game State/Feature Matrix
+        for tick in ticks:
```

Comment:
The logic here makes sense, what I think we want to do though is create a smaller function then use `.apply()` in pandas. Let's refactor along those lines.

Reply:
I have updated the function to use `.apply()`